Add Validation Tests, clear(), and reset()

LionOps opened the pull request:
Fixes https://github.com/yola/ampersand-multifield-view/issues/7
- `tests` is now an option that is array of functions that run validations on the multifield's `value`.
- `clear` can be called to clear all inputs.
- `reset` can be called to reset form to the initial state.

Commit:
Add custom validation with tests

Commit:
Add clear and reset functions

Commit:
Disable validation after clear

Commit:
Update README

Commit:
Simplify the conditional

LionOps commented:
:bell:

euwest commented on ampersand-multifield-view.js:
```diff
@@ -73,9 +98,25 @@ var MultiFieldView = View.extend({
+    var errMsg = '';
+
+    this.tests.some(function(test) {
+      errMsg = test.call(this, this.value) || '';
```
why the empty string?

LionOps replied:
It's type coercion to enforce that `this.message` is always a string. If there is no error, the test returns undefined or another falsey value, but not necessarily an empty string.

euwest replied:
since you're already importing `lodash.find`, you could also do something like:

``` js
runTests: function() {
  this.message = find(this.tests, function(test) {
    return test.call(this, this.value);
  }, this) || '';

  return this.message;
}
```

it's fine as is though, if you don't like the above

euwest commented on ampersand-multifield-view.js:
```diff
@@ -4,12 +4,17 @@ var find = require('lodash.find');
   template: [
     '<div>',
       '<p class="label" data-hook="label"></p>',
+      '<div data-hook="multi-message-container" class="message message-below message-error">',
```
are the classes being used?

LionOps replied:
Not within this plugin, but they're the same classes provided in other field views. [They](https://github.com/AmpersandJS/ampersand-input-view/blob/master/ampersand-input-view.js) aren't used in those views either.

euwest commented on ampersand-multifield-view.js:
```diff
@@ -138,12 +181,36 @@ var MultiFieldView = View.extend({
+
+  clear: function() {
+    // This won't unset field values, but will disable validation
+    this.setValue({}, true);
```
if you're just doing this to skip validation, let's just `this.shouldValidate = false;`

LionOps replied:
Ampersand's clear, when implemented, calls `this.setValue(<empty>, true)` where `<empty>` is something like `''` or `false`.

Functionally, `this.shouldValidate = false` is the same, but I was trying to preserve how other field-views work in calling `setValue` in their `clear` functions. I'm not sure it's worth keeping. Thoughts?

euwest replied:
if this view actually had its own `value` to set [distinct from its childrens' values] i'd say that makes sense, but as is it's just masking the invocation's true purpose(evidenced by the comment). I'm voting for direct `shouldValidate` manipulation over similar function appearance.

euwest commented on ampersand-multifield-view.js:
```diff
@@ -138,12 +181,36 @@ var MultiFieldView = View.extend({
+
+    this.fields.forEach(function(field) {
+      if(field.clear) {
+        field.clear();
```
this is in case of nested multi-fields?

LionOps replied:
In the case the fields support `clear()`. A lot of ampersand's views do so, but it's not a requirement for a field to support....

euwest commented on ampersand-multifield-view.js:
```diff
@@ -73,9 +98,25 @@ var MultiFieldView = View.extend({
+
+    var valid = !this.runTests();
+
+    return fieldsValid && valid;
```
you can skip the intermediary and make this a bit more efficient; if a field isn't valid, no need to check if the multi-field is valid:

``` js
return fieldsValid && !this.runTests();
```

I wish I could think of a better name for `runTests`, but alas...

Commit:
Clean up code
No need for an intermediary variable, bypass setValue for merely setting
shouldValidate

euwest commented:
:+1:
